Add statistics about memory states to the MemoryStateEncoder

haved opened the pull request:
This PR counts the number of memory states, and what type of `MemoryNode` they correspond to, for the following:
 - Region arguments (Intra-procedural only)
 - Loads (including volatile)
 - Stores (including volatile)
 - Call entry merges

It also tracks the largest amount of memory states used by a single such entity.

Finally, it tracks how many of the memory states correspond to `MemoryNode`s that are marked as module escaping.

phate commented on jlm/llvm/opt/alias-analyses/MemoryStateEncoder.cpp:
```diff
@@ -15,11 +15,140 @@
+      if (!memoryNode->IsModuleEscaping())
+        numNonEscaped++;
+
+      if (dynamic_cast<const PointsToGraph::AllocaNode *>(memoryNode))
+        numAllocas++;
+      else if (dynamic_cast<const PointsToGraph::MallocNode *>(memoryNode))
+        numMallocs++;
+      else if (dynamic_cast<const PointsToGraph::DeltaNode *>(memoryNode))
+        numDeltas++;
+      else if (dynamic_cast<const PointsToGraph::ImportNode *>(memoryNode))
+        numImports++;
+      else if (dynamic_cast<const PointsToGraph::LambdaNode *>(memoryNode))
+        numLambdas++;
+      else if (dynamic_cast<const PointsToGraph::ExternalMemoryNode *>(memoryNode))
+        numExternal++;
+      else
+        JLM_UNREACHABLE("Unknown MemoryNode type");
```
I think there is also an is<> method for memory nodes defined.

haved replied:
I could not find one, so I added one

phate replied:
There is already one: https://github.com/phate/jlm/blob/342848731a725be9648cb99fb5370596a325e0bc/jlm/llvm/opt/alias-analyses/PointsToGraph.hpp#L535C1-L542C4

phate commented on jlm/llvm/opt/alias-analyses/MemoryStateEncoder.cpp:
```diff
@@ -15,11 +15,140 @@
+  // The number of entities that have been counted
+  uint64_t NumEntities = 0;
+
+  // Count of total memory states, divided by MemoryNode type
```
"total memory states, separated by MemoryNode type"

phate commented on jlm/llvm/opt/alias-analyses/MemoryStateEncoder.cpp:
```diff
@@ -872,6 +1088,10 @@ MemoryStateEncoder::EncodeGammaEntry(rvsdg::GammaNode & gammaNode)
   auto memoryNodes = Context_->GetModRefSummary().GetGammaEntryNodes(gammaNode);
 
+  // Count the memory state arguments once per subregion
+  for (size_t i = 0; i < gammaNode.nsubregions(); i++)
```
for (auto & subregion : gammaNode.Subregions())

haved replied:
Hahah, yes, I should have remembered from the last two times you pointed this out :)

phate reviewed (approved):
Three minor comments.

haved commented:
I fixed your comments, @phate

I added an `is<>` for PointsToGraph Nodes, as I could not find one.

I also replaced the surrounding uses of `for(size_t i = 0; i < gammaNode.nsubregions(); i++)`, while I was at it

phate reviewed (changes requested):
There was already an Is<>() method. See comment.
